influxdb3_server/src/http/v1.rs

```diff
@@ -768,3 +768,129 @@ impl Stream for QueryResponseStream {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::array::ArrayRef;
+    use arrow::datatypes::{DataType, Field, Schema};
+    use arrow::record_batch::RecordBatch;
+    use arrow_array::{Float64Array, Int64Array, StringArray, TimestampNanosecondArray};
+    use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+    use futures::stream::{self, StreamExt};
+    use serde_json::json;
+    use std::sync::Arc;
+
+    fn times(vals: &[i64]) -> ArrayRef {
+        Arc::new(TimestampNanosecondArray::from_iter_values(
+            vals.iter().cloned(),
+        ))
+    }
+
+    fn strs<T: AsRef<str>>(vals: &[Option<T>]) -> ArrayRef {
+        Arc::new(StringArray::from_iter(vals))
+    }
+
+    fn f64s(vals: &[Option<f64>]) -> ArrayRef {
+        Arc::new(Float64Array::from_iter(vals.iter()))
+    }
+
+    fn i64s(vals: &[Option<i64>]) -> ArrayRef {
+        Arc::new(Int64Array::from_iter(vals.iter().cloned()))
+    }
+
+    fn create_test_record_batch() -> RecordBatch {
+        let meta = serde_json::to_string(&json!({
+            "measurement_column_index": 0,
+            "tag_key_columns": [],
+        }))
+        .unwrap();
+        let schema = Arc::new(Schema::new_with_metadata(
+            vec![
+                Field::new("iox::measurement", DataType::Utf8, false),
+                Field::new(
+                    "time",
+                    DataType::Timestamp(TimeUnit::Nanosecond, None),
+                    false,
+                ),
+                Field::new("cpu", DataType::Utf8, true),
+                Field::new("device", DataType::Utf8, true),
+                Field::new("usage_idle", DataType::Float64, true),
+                Field::new("free", DataType::Int64, true),
+            ],
+            HashMap::from([("iox::influxql::group_key::metadata".to_owned(), meta)]),
+        ));
+        RecordBatch::try_new(
+            schema,
+            vec![
+                strs(&[Some("cpu"), Some("cpu"), Some("cpu"), Some("cpu")]),
+                times(&[
+                    1157082300000000000,
+                    1157082310000000000,
+                    1157082400000000000,
+                    1157082320000000000,
+                ]),
+                strs(&[Some("cpu0"), Some("cpu0"), Some("cpu1"), Some("cpu2")]),
+                strs(&[Some("disk1s1"), None, Some("disk1s1"), None]),
+                f64s(&[Some(99.1), Some(99.8), Some(99.2), Some(99.3)]),
+                i64s(&[None, Some(2133), Some(4110), Some(1995)]),
+            ],
+        )
+        .unwrap()
+    }
+
+    #[tokio::test]
+    async fn test_partial_flag() {
+        let batch = create_test_record_batch();
+        let schema = batch.schema();
+        let input_stream = stream::iter(vec![Ok(batch.clone())]);
```

Comment:
Thanks for your suggestion. I have updated the test to align with that. With this new test, we can see that my implementation is not correct.

Reply:
Yes, I think that in order to make this work, the streaming/buffering logic needs to be refactored/rethought a bit. It would be great if you are willing to take a stab at it!

I updated the original issue description to make a note of this.

Reply:
Hmm, perhaps `peekable` is not the way to go. I wonder if the streaming logic, i.e., in the `impl Stream for QueryResponseStream` just needs to be re-ordered.

Right now, it always checks if it can flush before polling the input stream, which means that it could flush the entire buffer before polling the input stream again, and which seems to be the cause of the problem we are up against here.

If, instead, we just polled the input before flushing, and then, when flushing, always leave _at most_ one chunk in the buffer. That way, there is always enough data in the buffer, when streaming chunks, to decide what `partial` should be.

The exception would be, when the input stream is done, and there are no more record batches, then you would want to flush the entire buffer.

Reply:
@hiltontj Thanks for your feedback and description, but I'm a bit lost concerning the correct implementation and what I need to refactor. I don't think I will find the right approach 😅

Reply:
@hiltontj  I've just pushed another unit test to verify the existence of the partial flag when the stream size is 1